### PR TITLE
Feature/get wilaya by baladyia name

### LINKED ISSUE #47: getWilayaByBaladyiaName
Given a baladyia `name` or `code` ( would be more unique, because imo there is two baladya with same name and different wilaya).

### PR DESCRIPTION
**Pull request by Redaloukil**

### CODE REVIEW

File: src/api/getWilayaByBaladyiaName.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    const baladyiaName = baladyia.toLowerCase();
+    const wilaya = data.find(
+      w => w.dairats.find(
+        (d) => {
+          if(d.baladyiats){
+            return d.baladyiats.find((b) => b.name.toLowerCase() === baladyiaName || b.name_en.toLowerCase() === baladyiaName || b.name_ar === baladyiaName);
+          }
+          return null;
```
Comment (Fcmam5) [inline src/api/getWilayaByBaladyiaName.js:30]: This is not needed
 
```suggestion
          return d.baladyiats.find((b) => b.name.toLowerCase() === baladyiaName || b.name_en.toLowerCase() === baladyiaName || b.name_ar === baladyiaName);
```

File: tests/api/getWilayaByBaladyiaName.test.js [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    expect(result).toBeNull();
+  });
+
+  it('should return null if the wilaya with the given daira is empty', () => {
+    const result = getWilayaByBaladyiaName(mockData)("ZRZR");
+
+    expect(result).toBeNull();
+  });
+
+  it('should return matching wilaya object with arabic name', () => {
```
Comment (Fcmam5) [inline tests/api/getWilayaByBaladyiaName.test.js]: Wrong test name

File: tests/api/getWilayaByBaladyiaName.test.js [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+  it('should return matching wilaya object with english name', () => {
+    const result = getWilayaByBaladyiaName(mockData)("EL MENIAA", ["mattricule", "name"]);
+    expect(result).toEqual({
+      mattricule: 47,
+      name: "Ghardaïa",
+    });
+  });
+
+  it('should return matching wilaya object with baladyia arabe name', () => {
```
Comment (Fcmam5) [inline tests/api/getWilayaByBaladyiaName.test.js]: Wrong test name/case

File: tests/api/getWilayaByBaladyiaName.test.js [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  it('should return a curried function that returns the data', () => {
+    const fn = getWilayaByBaladyiaName(mockData);
+
+    expect(typeof fn).toBe('function');
+  });
+
+  it('should return null if the wilaya with the given baladyia is not found', () => {
+    const result = getWilayaByBaladyiaName(mockData)("foo");
+    expect(result).toBeNull();
+  });
```
Comment (Fcmam5) [inline tests/api/getWilayaByBaladyiaName.test.js]: This case is already covered in (`should return null if the wilaya with the given daira is empty`)
```suggestion
```

File: tests/api/getWilayaByBaladyiaName.test.js [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+  it('should return matching wilaya object with arabic name', () => {
+    const result = getWilayaByBaladyiaName(mockData)("GHARDAIA", ["mattricule", "name"]);
+
+    expect(result).toEqual({
+      mattricule: 47,
+      name: "Ghardaïa",
+    });
+  });
+
+  it('should return matching wilaya object with english name', () => {
```
Comment (Fcmam5) [inline tests/api/getWilayaByBaladyiaName.test.js]: I'd suggest adding something like `EL MENIAA - EN` in the initial dataset to cover this branch, or we can remove this test case anyway

File: tests/api/getWilayaByBaladyiaName.test.js [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+      name: "Ghardaïa",
+    });
+  });
+
+  it('should return matching wilaya object with baladyia arabe name', () => {
+    const result = getWilayaByBaladyiaName(noBaladyiaMockData)("امنيعة");
+    expect(result).toBeNull();
+  });
+
+  it('should return matching wilaya object with baladyia arabe name', () => {
```
Comment (Fcmam5) [inline tests/api/getWilayaByBaladyiaName.test.js]:
```suggestion
  it('should return matching wilaya object for a baladyia arabic name', () => {
```

File: tests/api/getWilayaByBaladyiaName.test.js [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    const result = getWilayaByBaladyiaName(noBaladyiaMockData)("المنيعة", ["mattricule"]);
+    expect(result).toEqual({
+      mattricule: 12,
+    });
+  });
```
Comment (Fcmam5) [inline tests/api/getWilayaByBaladyiaName.test.js:106]: Also, please add a test case where your projection array is empty

**Review by Fcmam5 — COMMENTED** [review]
I left few comments, mainly about the tests.
Also please update the README file

File: src/api/getWilayaByBaladyiaName.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+   *
+   * //returns {name : "Oran"}
+   * getWilayaByBaladyiaName("ES SENIA")
+   *
+   * @param { string } baladyia name
+   * @param { String[] } projection a list of baladyia object attributes to keep
+   * @returns { Object | null } Returns wilaya's object, or null
+   */
+  (baladyia, projection) => {
+    if (!baladyia || baladyia.trim().length < 3) {
```
Comment (Fcmam5) [inline src/api/getWilayaByBaladyiaName.js]: Please extract this magic number

File: tests/api/getWilayaByBaladyiaName.test.js [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    const result = getWilayaByBaladyiaName(mockData)("fo");
+
+    expect(result).toBeNull();
+  });
+
+  it('should return matching wilaya object for baladyia name', () => {
+    const result = getWilayaByBaladyiaName(mockData)("GHARDAIA");
+
+    expect(result).toEqual(mockData[0]);
+  });
```
Comment (Fcmam5) [inline tests/api/getWilayaByBaladyiaName.test.js:83]:
```suggestion
  });
 
```

**Review by Fcmam5 — DISMISSED** [review]
Thank you Reda!

File: tests/api/getWilayaByBaladyiaName.test.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
```
Comment (Fcmam5) [inline tests/api/getWilayaByBaladyiaName.test.js]:
```suggestion

```

**Review by Fcmam5 — APPROVED** [review]
